### DISCUSSION TITLE
controller boundary

### DISCUSSION THREAD

**Discussion by anda-ren** [star-whale/starwhale#800] [General]
Before k8s being imported into controller, controller doesn't need to understand SWRT, SWMP. If SWDS is refactored, controller doesn't need to understand SWDS either.

After k8s being imported into controller, controller take some responsibilites of the legacy agent , so that k8s could understand starwhale jobs. So controller need to understand SWRT, SWMP.

What is a clearer boundary for controller in the future?

**Comment by anda-ren** [comment]
I think controller has two main parts:

1. Starwhale remote instance
2. Distributed job scheduler

part 1 could be coupled with starwhale artifacts (SWDS,SWMP,SWRT)
part 2 should be decoupled with starwhale artifacts

**Reply by xuchuan** [reply]
First let me explain my understanding of the word "understand", correct me if I am wrong. I guess "understand" means that the controller should read information from the internal content of SWDS, SWMP, and SWRT and make some actions based on it. However, I do not see any difference between k8s and the legacy controller. The work that used to be done by the legacy agent can be accomplished in k8s pods or sidecars. I guess it is not necessary to move them into the controller. So would you please provide more concrete examples about the question?
